os/arch/arm/src/imxrt/imxrt_qtmr.c

```diff
@@ -1,563 +1,999 @@
+	priv->arg = arg;
+
+	if (priv->callback != NULL) {
+		imxrt_qtmr_setisr(priv->qtmr, imxrt_qtmr_handler, priv);
+		imxrt_qtmr_enableinterrupts(priv->qtmr->base, priv->channel, kQTMR_Compare1InterruptEnable);
+	} else {
+		imxrt_qtmr_disableinterrupts(priv->qtmr->base, priv->channel, kQTMR_Compare1InterruptEnable);
+		imxrt_qtmr_setisr(priv->qtmr, NULL, NULL);
+	}
+
+	irqrestore(flags);
+}
+
+static int imxrt_qtmr_settimeout(struct timer_lowerhalf_s *lower, uint32_t timeout)
+{
+	uint32_t freq;
+	uint64_t ticks;
+	struct imxrt_qtmr_lowerhalf_s *priv = (struct imxrt_qtmr_lowerhalf_s *)lower;
+
+	if (priv->started) {
+		return -EPERM;
+	}
+
+	freq = imxrt_clock_getperclkfreq();
+
+	ticks = ((uint64_t) freq * (uint64_t) timeout) / USEC_PER_SEC;
+	tmrdbg("freq:%d ticks:%llu, timeout = %lu\n", freq, ticks, timeout);
+
+	if (ticks < USHRT_MAX) {
+		imxrt_qtmr_settimerperiod(priv->qtmr->base, priv->channel, ticks);
+	} else {
+		tmrdbg("Ticks [%d] exceeds limit for this 16-bit QTMR timer counter [limit = %u]\n", ticks, USHRT_MAX);
+		return -EINVAL;
+	}
+
+	imxrt_qtmr_settimerperiod(priv->qtmr->base, priv->channel, ticks);
+
+	return OK;
+}
+
+static int imxrt_qtmr_getstatus(struct timer_lowerhalf_s *lower, struct timer_status_s *status)
+{
+	struct imxrt_qtmr_lowerhalf_s *priv = (struct imxrt_qtmr_lowerhalf_s *)lower;
+	uint32_t ticks, timeout_ticks, cur_ticks;
+	uint32_t freq;
+
+	freq = imxrt_clock_getperclkfreq();
+
+	/* Return the status bit */
+
+	status->flags = 0;
+	if (priv->started) {
+		status->flags |= TCFLAGS_ACTIVE;
+	}
+
+	if (priv->callback) {
+		status->flags |= TCFLAGS_HANDLER;
+	}
+
+	/* Return the actual timeout in microseconds */
+	timeout_ticks = imxrt_qtmr_gettimerperiod(priv->qtmr->base, priv->channel);
+	status->timeout = ((uint64_t)USEC_PER_SEC * (uint64_t)timeout_ticks) / freq;
+
+	/* Get the time remaining until the timer expires (in microseconds). */
+	cur_ticks = imxrt_qtmr_getcurrenttimercount(priv->qtmr->base, priv->channel);
+	ticks = timeout_ticks - cur_ticks;
+	status->timeleft = ((uint64_t)USEC_PER_SEC * (uint64_t)ticks) / freq;
+
+	return OK;
+}
+
+static int imxrt_qtmr_ioctl(struct timer_lowerhalf_s *lower, int cmd, unsigned long arg)
+{
+	return OK;
+}
+
+static int imxrt_qtmr_stop(struct timer_lowerhalf_s *lower)
+{
+	struct imxrt_qtmr_lowerhalf_s *priv = (struct imxrt_qtmr_lowerhalf_s *)lower;
+	tmrvdbg("Stop timer [%d] channel [%d]\n", priv->tmr, priv->channel);
+
+	if (priv->started) {
+		imxrt_qtmr_disableinterrupts(priv->qtmr->base, priv->channel, kQTMR_Compare1InterruptEnable);
+
+		imxrt_qtmr_stoptimer(priv->qtmr->base, priv->channel);
+
+		if (active[priv->tmr] == 1) {
+			imxrt_qtmr_setisr(priv->qtmr, NULL, NULL);
+		}
+		priv->started = false;
+
+		busy_status[priv->tmr][priv->channel] = false;
+
+		active[priv->tmr] -= 1;
+		tmrvdbg("QTMR Timer %d channel %d is stopped\n", priv->tmr, priv->channel);
+		return OK;
+	}
+
+	/* Return ENODEV to indicate that the timer was not running */
+
```

Comment:
We can remove these extra lines between comment and code

Reply:
ok